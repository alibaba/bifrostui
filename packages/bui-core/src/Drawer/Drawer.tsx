import clsx from 'clsx';
import React, { useState } from 'react';
import Modal from '../Modal';
import Slide from '../Slide';
import { DrawerProps } from './Drawer.types';
import './Drawer.less';

const prefixCls = 'bui-drawer';

const oppositeDirection = {
  left: 'right',
  right: 'left',
  top: 'down',
  bottom: 'up',
} as const;

const Drawer = React.forwardRef<HTMLDivElement, DrawerProps>((props, ref) => {
  const {
    anchor: anchorProp,
    container,
    children,
    className,
    hideBackdrop,
    onClose,
    SlideProps,
    open,
    disablePortal,
    contentProps,
    transitionDuration,
    ...others
  } = props;
  const anchorInvariant = oppositeDirection[anchorProp];
  const [keepMounted, setKeepMounted] = useState(open);

  return (
    <Modal
      container={container}
      open={open}
      onClose={onClose}
      hideBackdrop={hideBackdrop}
      ref={ref}
      className={clsx(prefixCls, className)}
      disablePortal={disablePortal}
      keepMounted={keepMounted}
      {...others}
    >
      <Slide
        in={open}
        direction={anchorInvariant}
        timeout={transitionDuration}
        onEnter={() => setKeepMounted(true)}
        onExited={() => setKeepMounted(false)}
        {...SlideProps}
      >
        <div
          {...contentProps}
          className={clsx(
            `${prefixCls}-content`,
            `${prefixCls}-content-${anchorProp}`,
            contentProps?.className,
          )}
        >
          {children}
        </div>
      </Slide>
    </Modal>
  );
});

Drawer.displayName = 'BuiDrawer';
Drawer.defaultProps = {
  anchor: 'bottom',
  hideBackdrop: false,
  open: false,
  disablePortal: false,
};

export default Drawer;
