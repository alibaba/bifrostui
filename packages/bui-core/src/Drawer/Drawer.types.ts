import React from 'react';
import { ModalProps } from '../Modal/Modal.types';
import { SlideProps } from '../Slide/Slide.types';
import { TransitionProps } from '../Transition';

export interface DrawerProps extends ModalProps {
  /**
   * Drawer出现的位置 'left' | 'top' | 'right' | 'bottom'
   * @default 'bottom'
   */
  anchor?: 'left' | 'top' | 'right' | 'bottom';
  /**
   * 是否显示
   * @default false
   */
  open?: boolean;
  /**
   * 是否隐藏遮罩层
   * @default false
   */
  hideBackdrop?: boolean;
  /**
   * 抽屉挂载的节点
   * @default 页面根节点
   */
  container?: ModalProps['container'];
  /**
   * Slide组件的Props
   */
  SlideProps?: Partial<SlideProps>;
  /**
   * 动态效果延时设置
   * @default {
   *    enter: duration.enteringScreen, // 225
   *    exit: duration.leavingScreen,  // 195
   *  }
   */
  transitionDuration?: TransitionProps['timeout'];
  /**
   * 抽屉关闭回调事件
   */
  onClose?: ModalProps['onClose'];
  /**
   * 是否禁用Portal
   * @default false
   */
  disablePortal?: boolean;
  /**
   * 抽屉内容节点的Props
   */
  contentProps?: React.HTMLAttributes<HTMLDivElement>;
}
