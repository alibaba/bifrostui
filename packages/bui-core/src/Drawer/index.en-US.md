---
group: feedback
name: Drawer drawer
---

# Drawer drawer

A floating panel that slides out from the edge of the screen.

## Code demonstration

### Basic usage

Control whether the drawer is displayed through 'open' and 'onClose'

```tsx
import { Button, Drawer, Stack } from '@bifrostui/react';
import React, { useState } from 'react';

export default () => {
  const [open, setOpen] = useState(false);
  return (
    <Stack direction="row">
      <Button
        onClick={() => {
          setOpen(true);
        }}
      >
        基础用法
      </Button>
      <Drawer
        open={open}
        onClose={() => {
          setOpen(false);
        }}
        contentProps={{ style: { lineHeight: '20vh', textAlign: 'center' } }}
      >
        抽屉正文
      </Drawer>
    </Stack>
  );
};
```

### Direction of appearance

By default, it appears from the bottom and supports controlling the appearance direction through setting 'anchor'. It appears from the bottom by default

```tsx
import { Button, Drawer, Stack } from '@bifrostui/react';
import React, { useState } from 'react';

export default () => {
  const [openDirection, setOpenDirection] = useState('');
  return (
    <>
      <Stack direction="row" spacing="8px">
        <Button onClick={() => setOpenDirection('bottom')}>底部弹出</Button>
        <Button onClick={() => setOpenDirection('top')}>顶部弹出</Button>
        <Button onClick={() => setOpenDirection('left')}>左侧弹出</Button>
        <Button onClick={() => setOpenDirection('right')}>右侧弹出</Button>
      </Stack>
      <Drawer
        anchor="bottom"
        open={openDirection === 'bottom'}
        onClose={() => setOpenDirection('')}
        contentProps={{ style: { lineHeight: '20vh', textAlign: 'center' } }}
      >
        底部弹出
      </Drawer>
      <Drawer
        anchor="top"
        open={openDirection === 'top'}
        onClose={() => setOpenDirection('')}
      >
        <div style={{ lineHeight: '20vh', textAlign: 'center' }}>顶部弹出</div>
      </Drawer>
      <Drawer
        anchor="left"
        open={openDirection === 'left'}
        onClose={() => setOpenDirection('')}
        contentProps={{
          style: { width: '50vw', textAlign: 'center', paddingTop: '50vh' },
        }}
      >
        左侧弹出
      </Drawer>
      <Drawer
        anchor="right"
        open={openDirection === 'right'}
        onClose={() => setOpenDirection('')}
      >
        <Stack style={{ width: '50vw', height: '100%' }}>right</Stack>
      </Drawer>
    </>
  );
};
```

### Dynamic effect delay

```tsx
import { Button, Drawer, Stack } from '@bifrostui/react';
import React, { useState } from 'react';
export default () => {
  const [open, setOpen] = useState(false);
  return (
    <Stack direction="row">
      <Button onClick={() => setOpen(true)}>动态效果延时</Button>
      <Drawer
        transitionDuration={1000}
        open={open}
        onClose={() => setOpen(false)}
      >
        <div style={{ lineHeight: '20vh', textAlign: 'center' }}>
          动态效果延时
        </div>
      </Drawer>
    </Stack>
  );
};
```

### Specify container

By default, Drawer is rendered to the root node of the page, H5 is' document. body ', and the mini program is the' page node '.

```tsx
import { Button, Drawer, Stack } from '@bifrostui/react';
import React, { useState } from 'react';

export default () => {
  const [open, setOpen] = useState(false);
  return (
    <Stack id="draw-container">
      <Button
        onClick={() => {
          setOpen(true);
        }}
      >
        指定容器
      </Button>
      <Drawer
        container={document.getElementById('draw-container')}
        open={open}
        onClose={() => {
          setOpen(false);
        }}
      >
        <div
          className="content"
          style={{ lineHeight: '20vh', textAlign: 'center' }}
        >
          浮层dom在`draw-container`element中渲染
        </div>
      </Drawer>
    </Stack>
  );
};
```

### API

| attribute          | explain                               | type                                                        | Default value   |
| ------------------ | ------------------------------------- | ----------------------------------------------------------- | --------------- |
| anchor             | The position where the Drawer appears | `left` \|`top` \|`right` \|`bottom`                         | `text`          |
| open               | Is it displayed                       | boolean                                                     | false           |
| hideBackdrop       | Do you want to hide the mask layer    | boolean                                                     | false           |
| container          | Element nodes mounted on drawers      | ModalProps['container']                                     | Pages and Nodes |
| SlideProps         | Props for Slide Animation             | Partial<SlideProps\>                                        | -               |
| transitionDuration | Animation execution time              | number \|{ appear?: number, enter?: number, exit?: number } | -               |
| onClose            | Close drawer callback                 | ModalProps['onClose']                                       | -               |
| disablePortal      | Prohibit Portal Capability            | boolean                                                     | false           |
| contentProps       | Props of drawer content element nodes | React.HTMLAttributes<HTMLDivElement\>                       | -               |

Drawer inherits from Modal. Other properties can be found in the Modal API (/cores/modal? # API)
