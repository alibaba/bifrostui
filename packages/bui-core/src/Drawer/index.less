.bui-drawer {
  font-family: var(--bui-font-family);

  &-content {
    display: flex;
    flex-direction: column;
    height: 100%;
    flex: 1 0 auto;
    position: fixed;
    top: 0;
    outline: 0;
    background-color: var(--bui-color-bg-view);

    &-left {
      left: 0;
    }

    &-top {
      top: 0;
      left: 0;
      right: 0;
      height: auto;
      max-height: 100%;
    }

    &-right {
      right: 0;
    }

    &-bottom {
      bottom: 0;
      top: auto;
      left: 0;
      /* stylelint-disable-next-line declaration-block-no-redundant-longhand-properties */
      right: 0;
      height: auto;
      max-height: 100%;
    }
  }
}
