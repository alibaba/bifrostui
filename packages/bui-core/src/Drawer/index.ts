export { default, default as Drawer } from './Drawer';
export * from './Drawer.types';
