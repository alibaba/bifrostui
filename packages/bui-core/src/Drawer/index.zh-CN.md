---
group: 反馈
name: Drawer 抽屉
---

# Drawer 抽屉

屏幕边缘滑出的浮层面板。

## 代码演示

### 基本使用

通过 `open` 以及 `onClose` 控制抽屉是否展示

```tsx
import { Button, Drawer, Stack } from '@bifrostui/react';
import React, { useState } from 'react';

export default () => {
  const [open, setOpen] = useState(false);
  return (
    <Stack direction="row">
      <Button
        onClick={() => {
          setOpen(true);
        }}
      >
        基础用法
      </Button>
      <Drawer
        open={open}
        onClose={() => {
          setOpen(false);
        }}
        contentProps={{ style: { lineHeight: '20vh', textAlign: 'center' } }}
      >
        抽屉正文
      </Drawer>
    </Stack>
  );
};
```

### 出现方向

默认从底部出现、支持通过设置 `anchor` 控制出现方向，默认是从底部出现

```tsx
import { Button, Drawer, Stack } from '@bifrostui/react';
import React, { useState } from 'react';

export default () => {
  const [openDirection, setOpenDirection] = useState('');
  return (
    <>
      <Stack direction="row" spacing="8px">
        <Button onClick={() => setOpenDirection('bottom')}>底部弹出</Button>
        <Button onClick={() => setOpenDirection('top')}>顶部弹出</Button>
        <Button onClick={() => setOpenDirection('left')}>左侧弹出</Button>
        <Button onClick={() => setOpenDirection('right')}>右侧弹出</Button>
      </Stack>
      <Drawer
        anchor="bottom"
        open={openDirection === 'bottom'}
        onClose={() => setOpenDirection('')}
        contentProps={{ style: { lineHeight: '20vh', textAlign: 'center' } }}
      >
        底部弹出
      </Drawer>
      <Drawer
        anchor="top"
        open={openDirection === 'top'}
        onClose={() => setOpenDirection('')}
      >
        <div style={{ lineHeight: '20vh', textAlign: 'center' }}>顶部弹出</div>
      </Drawer>
      <Drawer
        anchor="left"
        open={openDirection === 'left'}
        onClose={() => setOpenDirection('')}
        contentProps={{
          style: { width: '50vw', textAlign: 'center', paddingTop: '50vh' },
        }}
      >
        左侧弹出
      </Drawer>
      <Drawer
        anchor="right"
        open={openDirection === 'right'}
        onClose={() => setOpenDirection('')}
      >
        <Stack style={{ width: '50vw', height: '100%' }}>right</Stack>
      </Drawer>
    </>
  );
};
```

### 动态效果延时

```tsx
import { Button, Drawer, Stack } from '@bifrostui/react';
import React, { useState } from 'react';
export default () => {
  const [open, setOpen] = useState(false);
  return (
    <Stack direction="row">
      <Button onClick={() => setOpen(true)}>动态效果延时</Button>
      <Drawer
        transitionDuration={1000}
        open={open}
        onClose={() => setOpen(false)}
      >
        <div style={{ lineHeight: '20vh', textAlign: 'center' }}>
          动态效果延时
        </div>
      </Drawer>
    </Stack>
  );
};
```

### 指定容器

默认将 Drawer 渲染到页面的根节点中，H5是`document.body`, 小程序是`page节点`。

```tsx
import { Button, Drawer, Stack } from '@bifrostui/react';
import React, { useState } from 'react';

export default () => {
  const [open, setOpen] = useState(false);
  return (
    <Stack id="draw-container">
      <Button
        onClick={() => {
          setOpen(true);
        }}
      >
        指定容器
      </Button>
      <Drawer
        container={document.getElementById('draw-container')}
        open={open}
        onClose={() => {
          setOpen(false);
        }}
      >
        <div
          className="content"
          style={{ lineHeight: '20vh', textAlign: 'center' }}
        >
          浮层dom在`draw-container`element中渲染
        </div>
      </Drawer>
    </Stack>
  );
};
```

| 属性               | 说明                     | 类型                                                         | 默认值     |
| ------------------ | ------------------------ | ------------------------------------------------------------ | ---------- |
| anchor             | Drawer 出现的位置        | `left` \| `top` \| `right` \| `bottom`                       | `text`     |
| open               | 是否显示                 | boolean                                                      | false      |
| hideBackdrop       | 是否隐藏遮罩层           | boolean                                                      | false      |
| container          | 抽屉挂载的元素节点       | ModalProps['container']                                      | 页面跟节点 |
| SlideProps         | Slide 动画的 props       | Partial<SlideProps\>                                         | -          |
| transitionDuration | 动画执行时间             | number \| { appear?: number, enter?: number, exit?: number } | -          |
| onClose            | 关闭抽屉回调             | ModalProps['onClose']                                        | -          |
| disablePortal      | 禁止 Portal能力          | boolean                                                      | false      |
| contentProps       | 抽屉内容元素节点的 props | React.HTMLAttributes<HTMLDivElement\>                        | -          |

`Drawer` 继承自 `Modal` 其他属性见 [Modal API](/cores/modal?#API)
