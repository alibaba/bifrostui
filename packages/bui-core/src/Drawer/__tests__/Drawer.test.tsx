import React from 'react';
import { act, isConformant, render, screen, userEvent, waitFor } from 'testing';
import Drawer from '../index';

describe('Drawer', () => {
  const rootClass = 'bui-drawer';
  beforeEach(() => {
    jest.useFakeTimers();
  });
  afterEach(() => {
    jest.clearAllTimers();
    jest.useRealTimers();
  });

  isConformant({
    Component: Drawer,
    displayName: 'BuiDrawer',
    className: rootClass,
    requiredProps: {
      open: true,
    },
    getTargetElement: () => {
      return document.querySelector('.bui-modal');
    },
  });
  it('renders correctly', async () => {
    const { container } = render(
      <Drawer open disablePortal>
        Drawer
      </Drawer>,
    );
    await act(async () => {
      await jest.runAllTimers();
    });
    expect(container.firstChild).toMatchSnapshot();
  });
  it('should open and close', async () => {
    const { rerender } = render(<Drawer>Drawer content</Drawer>);
    expect(screen.queryByText('Drawer content')).not.toBeInTheDocument();
    rerender(<Drawer open>Drawer content</Drawer>);
    expect(screen.getByText('Drawer content')).toBeInTheDocument();
  });

  it('should hide backdrop with hideBackdrop props', () => {
    render(
      <Drawer open hideBackdrop>
        Drawer content
      </Drawer>,
    );
    expect(document.querySelector('.bui-backdrop')).not.toBeInTheDocument();
  });

  it.each(['left', 'top', 'right', 'bottom'] as const)(
    'should render correctly with anchor props',
    (anchor) => {
      render(
        <Drawer open anchor={anchor}>
          Drawer
        </Drawer>,
      );
      expect(screen.getByText('Drawer')).toHaveClass(
        `${rootClass}-content-${anchor}`,
      );
    },
  );

  it('onclose should be called when closing drawer', () => {
    const onClose = jest.fn();
    render(
      <Drawer open onClose={onClose}>
        Drawer
      </Drawer>,
    );
    userEvent.click(screen.getByText('Drawer'));
    userEvent.click(document.querySelector('.bui-backdrop'));
    expect(onClose).toHaveBeenCalledTimes(1);
  });

  it('should delay the slide transition to complete using default theme values by default', async () => {
    render(<Drawer open>Drawer</Drawer>);
    await act(async () => {
      await jest.runAllTimers();
    });
    expect(screen.getByText('Drawer')).toHaveStyle({
      transition: 'transform 225ms cubic-bezier(0.0, 0, 0.2, 1) 0ms',
    });
  });

  it('after the slide transition exited the root dom can remove', async () => {
    const { rerender } = render(<Drawer open>Drawer</Drawer>);
    await act(async () => {
      await jest.runAllTimers();
    });
    rerender(<Drawer>Drawer</Drawer>);
    expect(screen.getByText('Drawer')).toBeInTheDocument();
    await act(async () => {
      await jest.runAllTimers();
    });
    await waitFor(() => {
      expect(screen.queryByText('Drawer')).not.toBeInTheDocument();
    });
  });

  it('should delay the slide transition to complete using custom theme values', async () => {
    render(
      <Drawer open transitionDuration={225}>
        Drawer
      </Drawer>,
    );
    await act(async () => {
      await jest.runAllTimers();
    });
    expect(screen.queryByText('Drawer')).toHaveStyle({
      transition: 'transform 225ms cubic-bezier(0.0, 0, 0.2, 1) 0ms',
    });
  });

  it('contentProps can set content element', () => {
    render(
      <Drawer open contentProps={{ className: 'test-content' }}>
        Drawer
      </Drawer>,
    );
    expect(screen.getByText('Drawer')).toHaveClass('test-content');
  });
});
