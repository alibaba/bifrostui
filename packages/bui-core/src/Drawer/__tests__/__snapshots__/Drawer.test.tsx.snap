// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drawer renders correctly 1`] = `
<div
  class="bui-modal bui-drawer"
>
  <div
    class="bui-backdrop bui-modal-backdrop"
    style="visibility: visible; transition: opacity 225ms cubic-bezier(0.0, 0, 0.2, 1) 0ms; opacity: 1;"
  />
  <div
    class="bui-drawer-content bui-drawer-content-bottom"
    style="visibility: visible; transition: transform 225ms cubic-bezier(0.0, 0, 0.2, 1) 0ms; transform: none;"
  >
    Drawer
  </div>
</div>
`;
